### PR TITLE
ellaborated docs on determinism in dataset ops

### PR DESCRIPTION
**Pull request by jackd**
Following on from [this discussion](https://github.com/tensorflow/tensorflow/issues/44491), `Dataset.map` docs imply setting `determinism=True` in `Dataset.map` will make the returned dataset deterministic (assuming the original is). This clarifies the behaviour is undefined if `map_func` is not pure (doc addition similar to that used in `interleave`).

### CODE REVIEW

File: tensorflow/python/data/ops/dataset_ops.py [hunk 1]
```diff
@@ -1774,6 +1774,11 @@ def map(self, map_func, num_parallel_calls=None, deterministic=None):
     in general precludes the possibility of executing user-defined
     transformations in parallel (because of Python GIL).
 
+    The order of elements yielded by this transformation is
```
Comment (aaudiber) [inline tensorflow/python/data/ops/dataset_ops.py:1777]: Please clarify that this paragraph applies to cases where `num_parallel_calls > 1`. When `num_parallel_calls <= 1`, the order is deterministic as long as the `map_func` is deterministic (doesn't need to be a pure func and `deterministic=True` isn't needed).

**Review by aaudiber — COMMENTED** [review]
Looks good, thanks @jackd!

**Comment by jackd** [comment]
@aaudiber clarified. Is it worth adding an example? e.g.

```python

tf.random.set_seed(0)
dataset = Dataset.range(
    1, 6, output_type=tf.float32)  #  ==>[ 1., 2., 3., 4., 5. ]
dataset = dataset.map(lambda x: x + tf.random.uniform(()),
    num_parallel_calls=tf.data.AUTOTUNE,
    deterministic=True)
print(list(dataset.as_numpy_iterator()))
# [1.5166783, 2.1485605, 3.5400312, 4.0197573, 5.468353]
print(list(dataset.as_numpy_iterator()))
# [1.019757, 2.5400312, 3.4683528, 4.5166783, 5.1485605]
```

**Review by aaudiber — COMMENTED** [review]
IMO the text defining the determinism guarantees is enough and we don't need the example.

File: tensorflow/python/data/ops/dataset_ops.py [hunk 2]
```diff
@@ -1784,6 +1784,12 @@ def map(self, map_func, num_parallel_calls=None, deterministic=None):
     ...     num_parallel_calls=tf.data.AUTOTUNE,
     ...     deterministic=False)
 
+    If `num_parallel_calls > 1`, the order of elements yielded by this
```
[the remark this answers is not in the document]
Comment (jackd) [reply tensorflow/python/data/ops/dataset_ops.py]: ... In that case I need a better understanding of what `deterministic=True` means. That said, my understanding is "the order of elements yielded by this transformation is deterministic if `deterministic=True`" is True regardless of `num_parallel_calls`, so I've removed the leading condition.

[the remark this answers is not in the document]
Comment (aaudiber) [reply tensorflow/python/data/ops/dataset_ops.py]: Sorry, I was thinking about the values+order being deterministic, the original text was accurate (though it applied independently of the value of `num_parallel_calls`). Your understanding is correct.

I like the new text, it is clear and accurate, thanks!

**Review by aaudiber — APPROVED** [review]

**Pull request merged**